src/shiftedCompositeNormL2.jl

```diff
@@ -0,0 +1,223 @@
+"""
+mutable struct ShiftedCompositeNormL2{
+  R <: Real,
+  V0 <: Function,
+  V1 <: Function,
+  V2 <: AbstractMatrix{R},
+  V3 <: AbstractVector{R},
+} <: ShiftedCompositeProximableFunction
+  h::NormL2{R}
+  c!::V0
+  J!::V1
+  A::V2
+  Aᵧ::V2
+  b::V3
+  g::V3
+  res::V3
+  sol::V3
+  dsol::V3
+  p::V3
+  dp
+  function ShiftedCompositeNormL2(
+    λ::R,
+    c!::Function,
+    J!::Function,
+    A::AbstractMatrix{R},
+    b::AbstractVector{R},
+  ) where {R <: Real}
+    p = similar(b, A.n + A.m)
+    dp = similar(b, A.n + A.m)
+    g = similar(b)
+    res = similar(b)
+    sol = similar(b)
+    dsol = similar(b)
+    Aᵧ = SparseMatrixCOO(A.m,A.n + A.m, similar(A.rows,length(A.rows) + A.m), similar(A.cols, length(A.cols) + A.m), similar(A.vals,length(A.vals) + A.m))
+    if length(b) != size(A,1)
+      error("ShiftedCompositeNormL2: Wrong input dimensions, there should be as many constraints as rows in the Jacobian")
+    end
+    new{R,typeof(c!),typeof(J!),typeof(A),typeof(b)}(NormL2(λ),c!,J!,A,Aᵧ,b,g,res,sol,dsol,p,dp)
+  end
+end
+
+
+shifted(λ::R, c!::Function, J!::Function, A::AbstractMatrix{R}, b::AbstractVector{R}, xk :: AbstractVector{R}) where {R <: Real} = begin
+  c!(b,xk)
+  J!(A,xk)
+  ShiftedCompositeNormL2(λ,c!,J!,A,b)
+end
+
+shifted(
+  ψ::ShiftedCompositeNormL2{R, V0, V1, V2, V3},
+  xk::AbstractVector{R},
+) where {R <: Real, V0 <: Function, V1 <: Function, V2 <: AbstractMatrix{R}, V3<: AbstractVector{R}} = begin
+  b = similar(ψ.b)
+  ψ.c!(b,xk)
+  A = similar(ψ.A)
+  ψ.J!(A,xk)
+  ShiftedCompositeNormL2(ψ.h.lambda, ψ.c!, ψ.J!, A, b)
+end
+ 
+shifted(
+  ψ::CompositeNormL2{R, V0, V1, V2, V3},
+  xk::AbstractVector{R}
+) where {R <: Real, V0 <: Function, V1 <: Function, V2 <: AbstractMatrix{R}, V3 <: AbstractVector{R}} = begin
+  b = similar(ψ.b)
+  ψ.c!(b,xk)
+  A = copy(ψ.A)
+  ψ.J!(A,xk)
+  ShiftedCompositeNormL2(ψ.h.lambda, ψ.c!, ψ.J!, A, b)
+end
+
+fun_name(ψ::ShiftedCompositeNormL2) = "shifted L2 norm"
+fun_expr(ψ::ShiftedCompositeNormL2) = "t ↦ ‖c(xk) + J(xk)t‖₂"
+fun_params(ψ::ShiftedCompositeNormL2) = "c(xk) = $(ψ.b)\n" * " "^14 * "J(xk) = $(ψ.A)\n"
+
+function prox!(
+  y::AbstractVector{R},
+  ψ::ShiftedCompositeNormL2{R, V0, V1, V2, V3},
+  q::AbstractVector{R},
+  σ::R;
+  maxiter = 10000
+) where {R <: Real, V0 <: Function,V1 <:Function,V2 <: AbstractMatrix{R}, V3 <: AbstractVector{R}}
+  θ = R(0.8)
+  γ = R(0.0)
+  full_row_rank = true
+  # Initialize Aᵧ
+  ψ.Aᵧ.rows .= [ψ.A.rows;collect(eltype(ψ.A.rows),1:ψ.A.m)] 
+  ψ.Aᵧ.cols .= [ψ.A.cols;collect(eltype(ψ.A.cols),ψ.A.n+1:ψ.A.n + ψ.A.m)]
+  ψ.Aᵧ.vals .= [ψ.A.vals;zeros(eltype(ψ.A.vals),ψ.A.m)]
+
+  mul!(ψ.g, ψ.A, q)
+  ψ.g .+= ψ.b
+
+  spmat = qrm_spmat_init(ψ.Aᵧ; sym=false) # TODO: preallocate this
+  spfct = qrm_spfct_init(spmat) # TODO: preallocate this
+  qrm_set(spfct, "qrm_keeph", 0)
+
+  # Check interior convergence
+  qrm_analyse!(spmat, spfct; transp='t')
+  qrm_factorize!(spmat, spfct, transp='t')
+
+  qrm_solve!(spfct, ψ.g, ψ.p, transp='t')
```

Comment:
Shouldn't you check the rank of R before solving?

Reply:
Also, the code is a bit hard to follow because the notation does not match the paper. There's p here and p in the paper, but it's not the same p.

Reply:
It is to reduce the number of allocations. But we may change it if you find readibility more important: 
We need to solve 
```math
R^T R R^T p = g
```
For this, we could allocate three vectors $p_1, p_2,p$ and solve 
```math
\begin{align}
R^T p_1 &= g \\
R p_2 &= p_1\\
R^T p &= p_2
\end{align}
```
What I do instead is
```math
\begin{align}
R^T p &= g \\
R p_2 &= p\\
R^T p &= p_2
\end{align}
```
Where in fact $p_2$ is $q_k$ with the notation from the paper (In the code, it is called `ψ.sol` but I can rename it). This allows to save one allocation...